File: packages/ts-sdk/src/treedx/adapters/blobs.ts
import { binaryRequest, jsonRequest, segment, type TreeDxAdapterContext } from './common.js';
import type { BinaryBody } from '../types/index.js';

export class BlobsAdapter {
  constructor(private readonly context: TreeDxAdapterContext) {}
  read(repoId: string, input: unknown): Promise<unknown> { return jsonRequest(this.context.transport, 'POST', `/api/v1/repos/${segment(repoId)}/blobs/read`, input); }
  write(workspaceId: string, input: unknown): Promise<unknown> { return jsonRequest(this.context.transport, 'POST', `/api/v1/workspaces/${segment(workspaceId)}/blobs/write`, input); }
  delete(workspaceId: string, input: unknown): Promise<unknown> { return jsonRequest(this.context.transport, 'POST', `/api/v1/workspaces/${segment(workspaceId)}/blobs/delete`, input); }
  download(workspaceId: string, query?: Record<string, string | number | boolean | undefined>): Promise<unknown> { return jsonRequest(this.context.transport, 'GET', `/api/v1/workspaces/${segment(workspaceId)}/blobs/download`, undefined, query); }
  upload(workspaceId: string, body: BinaryBody, query?: Record<string, string | number | boolean | undefined>): Promise<unknown> { return binaryRequest(this.context.transport, 'PUT', `/api/v1/workspaces/${segment(workspaceId)}/blobs/upload`, body, query); }
  createMultipartUpload(workspaceId: string, input: unknown): Promise<unknown> { return jsonRequest(this.context.transport, 'POST', `/api/v1/workspaces/${segment(workspaceId)}/blobs/uploads`, input); }
  uploadPart(workspaceId: string, uploadId: string, partNumber: number, body: BinaryBody): Promise<unknown> { return binaryRequest(this.context.transport, 'PUT', `/api/v1/workspaces/${segment(workspaceId)}/blobs/uploads/${segment(uploadId)}/parts/${partNumber}`, body); }
  completeMultipartUpload(workspaceId: string, uploadId: string, input: unknown): Promise<unknown> { return jsonRequest(this.context.transport, 'POST', `/api/v1/workspaces/${segment(workspaceId)}/blobs/uploads/${segment(uploadId)}/complete`, input); }
  abortMultipartUpload(workspaceId: string, uploadId: string): Promise<unknown> { return jsonRequest(this.context.transport, 'DELETE', `/api/v1/workspaces/${segment(workspaceId)}/blobs/uploads/${segment(uploadId)}`); }
}

File: packages/elixir-sdk/lib/treedx_sdk/adapters/blobs.ex
defmodule TreeDxSdk.Blobs do
  @moduledoc false
  alias TreeDxSdk.Adapters.Common

  def read(client, repo_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/repos/" <> Common.segment(repo_id) <> "/blobs/read",
      body,
      %{}
    )
  end

  def write(client, workspace_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/blobs/write",
      body,
      %{}
    )
  end

  def delete(client, workspace_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/blobs/delete",
      body,
      %{}
    )
  end

  def download(client, workspace_id, query \\ %{}) do
    Common.json_request(
      client,
      :get,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/blobs/download",
      nil,
      query
    )
  end

  def upload(client, workspace_id, binary_body, query \\ %{}) do
    Common.binary_request(
      client,
      :put,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/blobs/upload",
      binary_body,
      query
    )
  end

  def create_multipart_upload(client, workspace_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/blobs/uploads",
      body,
      %{}
    )
  end

  def upload_part(client, workspace_id, upload_id, part_number, binary_body) do
    Common.binary_request(
      client,
      :put,
      "/api/v1/workspaces/" <>
        Common.segment(workspace_id) <>
        "/blobs/uploads/" <> Common.segment(upload_id) <> "/parts/" <> Common.segment(part_number),
      binary_body,
      %{}
    )
  end

  def complete_multipart_upload(client, workspace_id, upload_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <>
        Common.segment(workspace_id) <>
        "/blobs/uploads/" <> Common.segment(upload_id) <> "/complete",
      body,
      %{}
    )
  end

  def abort_multipart_upload(client, workspace_id, upload_id) do
    Common.json_request(
      client,
      :delete,
      "/api/v1/workspaces/" <>
        Common.segment(workspace_id) <> "/blobs/uploads/" <> Common.segment(upload_id),
      nil,
      %{}
    )
  end
end
